defect: algorithm RankByVotes should not mind the order in the tallysheet

alvarogarcia7 opened the pull request:
Example:
A-1
B-1
C-2
1 seat to allocate - there's no tie but a winner (C)

pau-minoves commented:
Hi Alvaro,

Thanks for the fix. This is definetly a bug, Intermediate ties should only produce a TIE when above the numberOfSeats threshold, regardless of the candidate order.

I'm reading your fix now. I want to see why we need to iterate candidates twice.

Meanwhile, I have some comments. You are mixing several changes on this pull request, which makes it dificult for me to accept. I think it would be better if you do pull requests from feature branches. It will also help you keep master clean.

I'll comment some of the changes individual changes:

- Ignore IDEA and infinitest conf are ok.
- Junit dependency is transitively pulled in from jbehave. Which is ok, because otherwise we pull in two diferent junit versions.
- Being this a library, I would like to keep it to 1.7. We are not using anyhting 1.8 specific and there is no need to limit the user to 1.8.
- I see you are adding ideas to the backlog. This is great but let's move use the issue system instead. This file was a notepad for me and it is not the way to go.

pau-minoves commented on pom.xml:
```diff
@@ -80,6 +89,13 @@
 		</dependency>
 
 		<dependency>
```
I know it is not written somewhere, but one of the design decisions of the library is to be as dependency light as posible. Why do we need guava?

alvarogarcia7 replied:
I'm using guava to get permutations for a property-based test that I didn't want to include in this pull request.

should fit into the next one

pau-minoves commented on src/main/java/org/jseats/model/methods/RankMethod.java:
```diff
@@ -1,17 +1,13 @@
-import org.jseats.model.Candidate;
-import org.jseats.model.InmutableTally;
-import org.jseats.model.Result;
+import org.jseats.model.*;
```
Please try to stick to the code format. I'm using default eclipse formatter.

pau-minoves commented on src/test/java/org/jseats/model/methods/RankMethodShould.java:
```diff
@@ -0,0 +1,165 @@
+
+	@Test
+	//Rank should not mind the order in the tallysheet
+	public void not_mind_the_order() throws Exception {
```
and to java notation :)

alvarogarcia7 replied:
Regarding java notation,

I try to stick to java notation on production code and test code except for
test method names. E.g,

notMindTheOrderOnTheTallySheetGivenAnEvenAmountOfCandidates vs
not_mind_the_order_on_the_tally_sheet_given_an_even_amount_of_candidates

Which one do you think is clearer?

to support my choice, see
http://programmers.stackexchange.com/questions/237561/naming-test-methods-in-java
(only answer)

also see this article on snake_case vs camelCase:
http://www.cs.kent.edu/~jmaletic/papers/ICPC2010-CamelCaseUnderScoreClouds.pdf

they conclude that novices understand better snake_case but with training
or experience the differences almost disappear

thanks in advance,

pau-minoves commented:
I pulled your changes into #bugfix/candidate-order-on-ties

alvarogarcia7 commented:
Regarding looping twice on the candidates, I do the first loop the exhaust the candidates that can be ranked without a tie, then deal with the rest (including tieBreaker).

You're right. I'd rather sort the options based on votes and properties, then loop picking the right ones. I guess the implementation I chose is not clear enough.

On a future commit I may rewrite the algorithm with the cited strategy (worst cost is O(numberOfCandidates+numberOfCandidates*log(numberOfCandidates) + numberOfSeats) )

Also, I guess the algorithm's cost has to be increased because yours was stopping on a tie (ie, search) and mine needs to get to the end (ie, exhaustive search)

In any case, yours is worst case O(numberOfCandidates+numberOfSeats*numberOfCandidates) and mine too [citation needed].

alvarogarcia7 commented:
I'm using java8 to be able to forEach,
removed comments from backlog,
created a tallybuilder (to ease the creation of tallies),
fixed imports (as you asked)

for more details, please refer to the commits
